uips/uip-4.md

```diff
@@ -0,0 +1,102 @@
+
+## Motivation
+
+DAGSync is a graph-aware fast syncing algorithm. A client, upon detecting a single transaction involving them, can check that outputs visible to them are spent or not. If the output is unspent, then they have identified a live note they can potentially spend in the future, else if the output is unspent, they can continue the process forwards in the transaction graph, until they reach unspent notes.
+
+The design of Penumbra currently does not allow traversal backwards through the transaction graph, only forwards. A `Spend` intentionally does not reveal the note being spent, only the nullifier that is revealed. By including on the `Spend` an encrypted reference back to the note commitment being spent, such that only the note owner can view it, we enable DAGSync clients to efficienctly reconstruct the transaction history both backwards and forwards.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Modification to `SpendBody`
+
+The `Spend` action will be augmented with an additional field `backref_commitment` on the `SpendBody`:
+
+```
+message SpendBody {
+    // A commitment to the value of the input note.
+    penumbra.core.asset.v1.BalanceCommitment balance_commitment = 1;
+    // The nullifier of the input note.
+    penumbra.core.component.sct.v1.Nullifier nullifier = 6;
+    // The randomized validating key for the spend authorization signature.
+    penumbra.crypto.decaf377_rdsa.v1.SpendVerificationKey rk = 4;
+    // NEW: An encryption of the commitment of the input note to the sender's OVK.
+    optional bytes backref_commitment = 7;
+}
+```
+
+Clients MAY populate the `backref_commitment` field with the encrypted note commitment corresponding to the note they are spending. Initially the field will be optional, allowing for a phased adoption period such that clients have time to implement support for `Spend` backreferences.
+
+### Encryption of Spend Backreference
+
+The `backref_commitment` should be encrypted using `ChaCha20-Poly1305`. [RFC-8349](https://datatracker.ietf.org/doc/rfc8439/) specifies that (key, nonce) pairs MUST NOT be reused.
+
+We derive a symmetric key $k$ from the `OutgoingViewingKey` $ovk$ using the BLAKE2b-512 hash function and personalization string `"Penumbra_Backref"`:
+
+```
+k = BLAKE2b-512("Penumbra_Backref", ovk)
+```
+
+One advantage of using a single key is that we can scan all spends using this key without having to do key derivation each time.
+
+A random 16-byte nonce $n$ will be generated and provided as a prefix in the `backref_commitment` field.
```

Comment:
How is the nonce generated? From an RNG? In that case, does this change the `Spend` APIs? (Unlike `Output`s, which already do encryption, I don't think `Spend`s currently do, would this change that?)

Since the `Spend` is supposed to be deterministically generated from the `SpendPlan`, do we need to add the nonce there too?

Or, could we (should we?) avoid needing to create a new nonce by using the (truncation of the) nullifier as the nonce? Saving 16 bytes is non-essential, but avoiding having to do extra code changes to the transaction planning system to maintain determinism seems like a win.

Reply:
I think using the first 16 bytes of the nullifier is fine, because each commitment will be guaranteed to have a different nullifier, so there's no nonce reuse between different commitments ; this avoids needing to modify the SpendPlan

Reply:
(This is why, looking backwards, and forwards to when we do end up needing to modify a plan, we should have instead, imo, just included a seed field in a given plan, which could then be used to generate any randomness needed by the plan with a CPRNG)